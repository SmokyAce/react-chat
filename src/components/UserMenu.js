import React, { lazy, Suspense } from 'react';
import PropTypes from 'prop-types';
import IconButton from '@material-ui/core/IconButton';
import AccountCircle from '@material-ui/icons/AccountCircle';
import Menu from '@material-ui/core/Menu';
import MenuItem from '@material-ui/core/MenuItem';
import Spinner from './Spinner';

const UserProfile = lazy(() => import('./UserProfile'));

class UserMenu extends React.Component {
  static propTypes = {
    user: PropTypes.shape({
      firstName: PropTypes.string,
      lastName: PropTypes.string,
      username: PropTypes.string.isRequired,
    }).isRequired,
    isConnected: PropTypes.bool.isRequired,
    actions: PropTypes.shape({
      logout: PropTypes.func.isRequired,
      editUser: PropTypes.func.isRequired,
    }).isRequired,
  };

  state = {
    anchorEl: null,
    editProfile: false,
  };

  handleClick = (event) => {
    this.setState({ anchorEl: event.currentTarget });
  };

  handleClose = () => {
    this.setState({ anchorEl: null, editProfile: false });
  };

  handleOpenProfile = () => {
    this.setState({
      anchorEl: null,
      editProfile: true,
    });
  };

  render() {
    const { actions, user, isConnected } = this.props;
    const { anchorEl, editProfile } = this.state;

    return (
      <>
        <IconButton
          aria-owns={anchorEl ? 'simple-menu' : null}
          aria-haspopup='true'
          onClick={this.handleClick}
          color='inherit'
          disabled={!isConnected}
        >
          <AccountCircle />
        </IconButton>
        <Menu
          id='simple-menu'
          anchorEl={anchorEl}
          open={Boolean(anchorEl)}
          onClose={this.handleClose}
        >
          <MenuItem onClick={this.handleOpenProfile}>Edit Profile</MenuItem>
          <MenuItem onClick={actions.logout}>Logout</MenuItem>
        </Menu>
        <Suspense fallback={<Spinner />}>
          <UserProfile
            open={editProfile}
            onClose={this.handleClose}
            editUser={actions.editUser}
            user={user}
          />
        </Suspense>
      </>
    );
  }
}

export default UserMenu;
