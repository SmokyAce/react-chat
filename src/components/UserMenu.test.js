/* eslint-env jest */
import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import UserMenu from './UserMenu';

const mockProps = {
  isConnected: false,
  user: {
    username: 'test',
  },
  actions: {
    logout: jest.fn(),
    editUser: jest.fn(),
  },
};

describe('<UserMenu />', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<UserMenu {...mockProps} />, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  it('renders correctly', () => {
    const tree = renderer.create(<UserMenu {...mockProps} />).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('renders correctly is connected', () => {
    const tree = renderer.create(<UserMenu {...mockProps} isConnected />).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
