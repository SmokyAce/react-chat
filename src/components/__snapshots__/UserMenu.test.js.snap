// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UserMenu /> renders correctly 1`] = `
<button
  aria-haspopup="true"
  aria-owns={null}
  className="MuiButtonBase-root-38 MuiButtonBase-disabled-39 MuiIconButton-root-32 MuiIconButton-colorInherit-33 MuiIconButton-disabled-36"
  disabled={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  tabIndex="-1"
  type="button"
>
  <span
    className="MuiIconButton-label-37"
  >
    <svg
      aria-hidden="true"
      className="MuiSvgIcon-root-41"
      focusable="false"
      role="presentation"
      viewBox="0 0 24 24"
    >
      <path
        d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 3c1.66 0 3 1.34 3 3s-1.34 3-3 3-3-1.34-3-3 1.34-3 3-3zm0 14.2c-2.5 0-4.71-1.28-6-3.22.03-1.99 4-3.08 6-3.08 1.99 0 5.97 1.09 6 3.08-1.29 1.94-3.5 3.22-6 3.22z"
      />
      <path
        d="M0 0h24v24H0z"
        fill="none"
      />
    </svg>
  </span>
</button>
`;

exports[`<UserMenu /> renders correctly is connected 1`] = `
<button
  aria-haspopup="true"
  aria-owns={null}
  className="MuiButtonBase-root-38 MuiIconButton-root-32 MuiIconButton-colorInherit-33"
  disabled={false}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  tabIndex="0"
  type="button"
>
  <span
    className="MuiIconButton-label-37"
  >
    <svg
      aria-hidden="true"
      className="MuiSvgIcon-root-41"
      focusable="false"
      role="presentation"
      viewBox="0 0 24 24"
    >
      <path
        d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 3c1.66 0 3 1.34 3 3s-1.34 3-3 3-3-1.34-3-3 1.34-3 3-3zm0 14.2c-2.5 0-4.71-1.28-6-3.22.03-1.99 4-3.08 6-3.08 1.99 0 5.97 1.09 6 3.08-1.29 1.94-3.5 3.22-6 3.22z"
      />
      <path
        d="M0 0h24v24H0z"
        fill="none"
      />
    </svg>
  </span>
  <span
    className="MuiTouchRipple-root-67"
  />
</button>
`;
